### ISSUE TITLE
[question] are RPATHs they removed by default? 

### ISSUE THREAD

**Issue by fulara** [conan-io/conan#6656]
I am seeing this on both Centos7 and ubuntu. 
Although in Centos it is: `RUNPATH` rather that `RPATH` that stays in binary, these are slightly different things but probably both should be removed.

I am almost sure that this is a bug, but raising this as a question anyway, just in case I am confused.

reading through: <https://docs.conan.io/en/latest/howtos/manage_shared_libraries/rpaths.html> it seem to state that RPATHs are disabled by default and one needs to add `KEEP_RPATH` to have it.

My suspicion is that this is caused because I am not using `cmake.install` in my cases. Please confirm if thats the case.

See below example:

liba:
```
[fulara@localhost liba]$ cat conanfile.py 
from conans import ConanFile, tools, CMake

class Conan(ConanFile):
    name = "liba"
    
    exports_sources = "include/*", "src/*", "CMakeLists.txt",
    options = {"shared": [True, False]}
    default_options = {"shared": True}
    generators = "cmake"

    def build(self):
        cmake = CMake(self)
        # same as cmake.configure(source_folder=self.source_folder, build_folder=self.build_folder)
        cmake.configure()
        cmake.build()
        # Build the "install" target, defining CMAKE_INSTALL_PREFIX to self.package_folder
        #cmake.install()

    def package(self):
        self.copy("include/*");
        self.copy("lib/*");

    def package_info(self):
        self.cpp_info.libs = tools.collect_libs(self)

[fulara@localhost liba]$ cat CMakeLists.txt 
project (cpp)

include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
conan_basic_setup()

add_library (cpp src/h.cpp)


[fulara@localhost liba]$ cat include/h.h 
int abc();
[fulara@localhost liba]$ cat src/h.cpp 
int abc() {
	return 42;
}

#built with:
conan create . 1.0@
```

bin:
```
[fulara@localhost bin]$ cat CMakeLists.txt 
project (bin)

include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
conan_basic_setup(TARGETS)

add_executable (bin src/main.cpp)
target_link_libraries(bin CONAN_PKG::liba)


[fulara@localhost bin]$ cat conanfile.py 
from conans import ConanFile, CMake

class Conan(ConanFile):
    name = "bin"

    exports_sources = "src/*", "CMakeLists.txt",
    generators = "cmake"

    requires = "liba/1.0@",

    def build(self):
        cmake = CMake(self)
        cmake.configure()
        cmake.build()

    def package(self):
        self.copy("bin/*");
[fulara@localhost bin]$ cat src/main.cpp 
#include "h.h"

int main() {
	return abc();
}

#built with:
conan create . 1.0@
```
Then Searching for RPATH:

```
cd package_dir/bin..
readelf -d bin | head -20 | grep -i RPATH
 0x000000000000000f (RPATH)              Library rpath: [/home/fulara/.conan/data/liba/1.0/_/_/package/2a623e3082a38f90cd2c3d12081161412de331b0/lib]
```

**Comment by jgsogo** [comment]
Hi, @fulara

RPATHs are removed only for Macos. This is the relevant snippet in [CMake generated code](https://github.com/conan-io/conan/blob/develop/conans/client/generators/cmake_common.py):

```cmake
macro(conan_set_rpath)
    conan_message(STATUS "Conan: Adjusting default RPATHs Conan policies")
    if(APPLE)
        # https://cmake.org/Wiki/CMake_RPATH_handling
        # CONAN GUIDE: All generated libraries should have the id and dependencies to other
        # dylibs without path, just the name, EX:
        # libMyLib1.dylib:
        #     libMyLib1.dylib (compatibility version 0.0.0, current version 0.0.0)
        #     libMyLib0.dylib (compatibility version 0.0.0, current version 0.0.0)
        #     /usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 120.0.0)
        #     /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1197.1.1)
        set(CMAKE_SKIP_RPATH 1)  # AVOID RPATH FOR *.dylib, ALL LIBS BETWEEN THEM AND THE EXE
                                    # SHOULD BE ON THE LINKER RESOLVER PATH (./ IS ONE OF THEM)
        # Policy CMP0068
        # We want the old behavior, in CMake >= 3.9 CMAKE_SKIP_RPATH won't affect the install_name in OSX
        set(CMAKE_INSTALL_NAME_DIR "")
    endif()
endmacro()
```

Here it is the docs about CMake behavior, for reference: https://gitlab.kitware.com/cmake/community/-/wikis/doc/cmake/RPATH-handling#default-rpath-settings

> By default if you don't change any RPATH related settings, CMake will link the executables and shared libraries with full RPATH to all used libraries in the build tree. When installing, it will clear the RPATH of these targets so they are installed with an empty RPATH

> `CMAKE_SKIP_RPATH`, if it is enabled all other RPATH related options are ignored, no RPATH is built into anything

so, yes, if you don't clean the rpaths and don't use `cmake.install`, rpaths will be there.

**Comment by fulara** [comment]
Okay, I take it as most proper way to use `conan`+`cmake` then would be to use `cmake.install` in the package step then, rather than what I am doing now which is a self.copy.

**Comment by jgsogo** [comment]
Anyway, there are a lot of recipes out there using `self.copy`, so, if you see any bug related to it or there is any improvement we can do in Conan to better handle this scenario let us know.

Right now, for example, I have one concern: if the rpath is encoded in a dependency it will look for the requirement in that folder (a folder in the Conan cache). Conan can create/download a different revision of the library that is stored in the same folder (same version, same packageID). Then the former binary will look for and find a wrong binary for its requirement... and according to the link above, in Unix, it looks first for `RPATH` before using `LD_LIBRARY_PATH`

**Comment by fulara** [comment]
Yes @jgsogo, thats actually a problem that I am dealing with right now, but it only happens on local dev boxes not on prod boxes so its lower priority for us, thats why I only got to look at this now. 
Because prod boxes always have different package storage path.

I think problem here is what I raised there: https://github.com/conan-io/conan/issues/6656 I havent had the chance to responded to memshared yet, Anyway what I mean by problem is that conan does not tell about the recommended way of connecting `cmake` + `conan` - in all packages at my company we were doing `self.copy` which I am now actively reworking.

But thats ofc documentation - conan behavior - not sure atm how it could be better.
